aten/src/ATen/native/cuda/TensorModeKernel.cu

```diff
@@ -0,0 +1,321 @@
+  // calculations to get an offset
+  scalar_t* data = self.data_ptr<scalar_t>();
+  for (int64_t i = 0; i < position.size(); i++) {
+    data += position[i] * ensure_nonempty_stride(self, i);
+  }
+
+  int64_t ndim = ensure_nonempty_dim(self.dim());
+  int64_t n_element = ensure_nonempty_size(self, ndim - 1);
+  THCThrustAllocator thrust_allocator(state);
+
+  // Wrap input data in Thrust device vector
+  thrust::device_ptr<scalar_t> vec_ptr = thrust::device_pointer_cast(data);
+  thrust::device_vector<scalar_t> iter(vec_ptr, vec_ptr + n_element);
+
+  // Fill sort_buffer with [0, 1, 2, ... n_element - 1]
+  thrust::sequence(
+      thrust::cuda::par(thrust_allocator).on(stream),
+      sort_buffer.begin(),
+      sort_buffer.end());
+
+  // Sort the input data. The original indices of the data are stored in
+  // sort_buffer
+  thrust::sort_by_key(
+      thrust::cuda::par(thrust_allocator).on(stream),
+      iter.begin(),
+      iter.end(),
+      sort_buffer.begin());
+
+  // Count # of unique elements via an inner product between adjacent elements.
+  // Add 1 if two neighboring element are not equal.
+  int unique = 1 +
+      thrust::inner_product(
+                   thrust::cuda::par(thrust_allocator).on(stream),
+                   iter.begin(),
+                   iter.end() - 1,
+                   iter.begin() + 1,
+                   0,
+                   thrust::plus<int>(),
+                   thrust::not_equal_to<scalar_t>());
+
+  // Count frequency of each element
+  thrust::device_vector<scalar_t> keys(unique);
+  thrust::device_vector<int> counts(unique);
+  thrust::reduce_by_key(
+      thrust::cuda::par(thrust_allocator).on(stream),
+      iter.begin(),
+      iter.end(),
+      thrust::constant_iterator<int>(1),
+      keys.begin(),
+      counts.begin());
+
+  // Find index of maximum count
+  auto it = thrust::max_element(
+      thrust::cuda::par(thrust_allocator).on(stream),
+      counts.begin(),
+      counts.end());
+  scalar_t mode = keys[it - counts.begin()];
+
+  // Find first index within which it occurs
+  auto position_iter = thrust::find(
+      thrust::cuda::par(thrust_allocator).on(stream),
+      iter.begin(),
+      iter.end(),
+      mode);
+
+  TORCH_INTERNAL_ASSERT(position_iter != iter.end());
+  int64_t index = sort_buffer[position_iter - iter.begin()];
+
+  // Place mode, index in output
+  scalar_t* values_data = values.data_ptr<scalar_t>();
+  int64_t* indices_data = indices.data_ptr<int64_t>();
+
+  for (int64_t i = 0; i < position.size(); i++) {
+    int64_t pos = position[i];
+    values_data += ensure_nonempty_stride(values, i) * pos;
+    indices_data += ensure_nonempty_stride(indices, i) * pos;
+  }
+
+  AT_CUDA_CHECK(cudaMemcpyAsync(
+      values_data, &mode, sizeof(scalar_t), cudaMemcpyHostToDevice, stream));
+  AT_CUDA_CHECK(cudaMemcpyAsync(
+      indices_data, &index, sizeof(scalar_t), cudaMemcpyHostToDevice, stream));
```

Comment:
This is dangerous because the copy might not have finished by the time `mode` and `index` go out of scope. You'll need an `AT_CUDA_CHECK(cudaStreamSynchronize(stream))` after.